### PR TITLE
Update watchman (and create all the packages it now depends on)

### LINKED ISSUE #156177: Watchman 4.9 is 4+ years old
###### Checklist

- [X] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [X] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: watchman

current version: 4.9.0
desired version: 2022.01.17.00

###### Notify maintainers

maintainers: @cstrahan 

###### Note for maintainers

Please tag this issue in your PR.

## Watchman is effectively versionless now
[From the watchman website](https://facebook.github.io/watchman/docs/install.html#binary-downloads-for-linux-macos-and-windows-beta):

> Watchman is continuously deployed as it passes our internal test validation inside Facebook and doesn’t use manually assigned or “approved” version numbers.

I'm not sure what this means when it comes to packaging in nix. The version strings are basically dates now, e.g.

```
watchman --version                                                                                                                                          ~/flexport/flexport
2022.01.17.00
```

I probably don't need the latest version, but switching to the homebrew installed version solved some mysterious bugs with dependent tools I was using (namely: `flow`).

### PR DESCRIPTION
**Pull request by kylesferrazza**
###### Description of changes

Updates watchman to the most recent release (closes #156177). Also brings in all of its dependencies.

Lots of help from https://github.com/NixOS/nixpkgs/pull/63097 (thanks @pierreis -- let me know if you don't want to be listed as a maintainer). Had to update most of the packages since that PR.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kylesferrazza** [comment]
`edencommon` does not seem to get tags for releases - please let me know if there's a preferred versioning method here rather than using the commit SHA.

**Comment by kylesferrazza** [comment]
Updated all versions from `2022.07.11.00` to `2022.07.25.00`.

**Comment by kylesferrazza** [comment]
@matthewbauer @LnL7 looks like you both have committed to `watchman` most recently -- can I get a review please?

**Comment by kylesferrazza** [comment]
Rebased onto latest `master`

**Comment by happenslol** [comment]
Any news on this? I'm very interested and I'm considering whether it makes sense to package this for myself until it gets merged.

File: pkgs/development/libraries/rsocket-cpp/default.nix
Comment (Atry) [inline pkgs/development/libraries/rsocket-cpp/default.nix:28]: Why does it depend on `fmt_8` while other packages depend on `fmt_9`? Would there be any conflict?

Comment (kylesferrazza) [reply pkgs/development/libraries/rsocket-cpp/default.nix:28]: Just a typo I made during rebasing. I'll update to fmt_9 and ensure it still builds and runs

**Comment by kylesferrazza** [comment]
Rebased onto `master` and updated `rsocket-cpp` to use `fmt_9`. Confirmed `watchman` builds and runs properly.

**Review by Atry — APPROVED** [review]
This PR looks good to me, but I don't have the merge permission.

File: pkgs/development/libraries/wangle/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/wangle/default.nix:22]: We don't need () here

File: pkgs/development/libraries/rsocket-cpp/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/rsocket-cpp/default.nix:24]: Why no tests?

File: pkgs/development/tools/watchman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/watchman/default.nix:21]:
```suggestion
    "-DWATCHMAN_STATE_DIR=/run/watchman"
```

File: pkgs/development/tools/watchman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/watchman/default.nix:4]:
```suggestion
, CoreServices
```

File: pkgs/development/tools/watchman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/watchman/default.nix:58]:
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [ CoreServices ];
```

File: pkgs/development/tools/watchman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/watchman/default.nix:72]: Why do we need the lock file twice? Can we fetchCrate instead?

File: pkgs/development/tools/watchman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/watchman/default.nix:84]:
```suggestion
    platforms   = platforms.unix;
```

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:18785]: We don't really sort all-packages that much but libraries should be near libraries and not applications.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by andrewhamon** [comment]
@kylesferrazza happy new year! Do you plan to pick this back up? If not, @SuperSandro2000's feedback seems relatively straightforward, so I wouldn't mind picking this up.

**Comment by kylesferrazza** [comment]
@andrewhamon happy new year to you as well!! I would appreciate if you had the time to implement the feedback, thanks. I can give you write access to my fork.

File: pkgs/top-level/all-packages.nix
Comment (andrewhamon) [reply pkgs/top-level/all-packages.nix:18785]: Hey @SuperSandro2000, maybe I'm blind but all these libraries appear to be near other libraries.

**Comment by andrewhamon** [comment]
Trying to get this building on my arm mac, I had to update the platforms on several of these packages, and now bumping into this error building fbthrift

```
error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
```

I see you tried to address this with this line: https://github.com/NixOS/nixpkgs/pull/181787/files#diff-2b78e1719b2b799aa459d40ebb853e1eca68a96078d25eb4be51bac9c895e152R23

Seems a bit off that 10.13 was working for you, given the error explicitly says 10.14.

**Comment by andrewhamon** [comment]
But if I change `"-DCMAKE_OSX_DEPLOYMENT_TARGET=10.13"` to `10.14` I can build on arm! Nice! No to address this feedback.

**Comment by andrewhamon** [comment]
Okay, merged master, as expected builds were broken. I'm upgrading everything to latest.

- `rsocket-cpp` was broken, but I don't think its a dependency in newer versions of these packages. Solution: d
- newer versions of these packages were broken when building against `folly`. I upgraded `folly` to latest as well - if this ends up breaking downstream packages I might try going the other direction and pinning everything in this PR to versions contemporary with the current version of `folly` in nixpkgs.

At this point, I have everything building _except_ watchman itself, which feels like decent progress.

File: pkgs/development/libraries/wangle/default.nix
Comment (andrewhamon) [reply pkgs/development/libraries/wangle/default.nix:22]: fixed

File: pkgs/development/tools/watchman/default.nix
Comment (andrewhamon) [reply pkgs/development/tools/watchman/default.nix]: Fixed

File: pkgs/development/tools/watchman/default.nix
Comment (andrewhamon) [reply pkgs/development/tools/watchman/default.nix]: Fixed

File: pkgs/development/tools/watchman/default.nix
Comment (andrewhamon) [reply pkgs/development/tools/watchman/default.nix]: Fixed

File: pkgs/development/tools/watchman/default.nix
Comment (andrewhamon) [reply pkgs/development/tools/watchman/default.nix]: Fixed, in multiple locations

**Comment by andrewhamon** [comment]
Alright, I have watchman building again on latest, I just had to update the cargo.lock.

My final quest will be to get to the bottom of why we need both `Cargo.lock` as well at the `cargo-lock.patch`

**Comment by andrewhamon** [comment]
@kylesferrazza in order to adhere to the contributor guidelines, I think I'm going to have to rewrite history to clean things up. My plan for that is to keep you as author, and add myself as a co-contributor in the footer of the commit [according to this guide.](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors#creating-co-authored-commits-on-the-command-line).

Let me know if that works for you, and also let me know if you'd prefer I open a new PR instead of force push to your branch.

**Comment by kylesferrazza** [comment]
> My plan for that is to keep you as author, and add myself as a co-contributor in the footer of the commit 
> Let me know if that works for you,

Go for it -- thanks for all your work here!

**Comment by andrewhamon** [comment]
Alright, last hurdle passed - no more need for the redundant `cargo-lock.patch`. I grepped the codebase for instances of `cargoRoot = ` and found some good examples to reference.

**Comment by andrewhamon** [comment]
Result of `nixpkgs-review pr 181787` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>buck</li>
    <li>edencommon</li>
    <li>fb303</li>
    <li>fbthrift</li>
    <li>fizz</li>
    <li>folly</li>
    <li>python310Packages.pywatchman</li>
    <li>python311Packages.pywatchman</li>
    <li>rs-git-fsmonitor</li>
    <li>wangle</li>
    <li>watchman</li>
    <li>wdt</li>
  </ul>
</details>

File: pkgs/development/tools/watchman/default.nix
Comment (andrewhamon) [reply pkgs/development/tools/watchman/default.nix]: Fixed

**Comment by kylesferrazza** [comment]
@andrewhamon should we rebase the last three commits in as fixups?

**Comment by andrewhamon** [comment]
Yep, just did that :)

**Comment by andrewhamon** [comment]
It seems like just about everything needs the `"-DCMAKE_OSX_DEPLOYMENT_TARGET=10.14"` fix

**Comment by andrewhamon** [comment]
I added it to watchman pre-emptively since it too depends on folly (which seems to be the source of the errors - deps that compile against folly headers). Hoping for all green this build 🤞

File: pkgs/development/tools/watchman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/watchman/default.nix:69]:
```suggestion
  postPatch = ''
```
Also can we reduce the scope of selected files easily?

File: pkgs/development/tools/watchman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/watchman/default.nix:31]: We can use ensureNewerSourcesForZipFilesHook instead

File: pkgs/development/libraries/fizz/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/fizz/default.nix:35]:
```suggestion
  cmakeFlags = [ "-Wno-dev" ]
    ++ lib.optionals stdenv.isDarwin [
```
nit

**Review by SuperSandro2000 — COMMENTED** [review]
I think we can merge this after those few comments.

**Comment by kylesferrazza** [comment]
Building locally, testing, then will push rebased branch

**Comment by andrewhamon** [comment]
So after doing some manual testing, I've learned two things:

- `patchShebangs .` doesn't seem to be necessary anymore.
- `-DWATCHMAN_STATE_DIR=/run/watchman` results in a broken-by-defualt experience, since `/run` isn't writable by non-root. The `--disable-statedir` option we were using before doesn't seem to work anymore, but it is functionally equivalent to using `/tmp` on NixOS (since that is what watchman ultimately falls back to). It _is_ functionally different on macOS, though, since macOS populates `$TMPDIR` with a private-per-user path on each boot.
  
  I propose `-DWATCHMAN_STATE_DIR=/tmp` as the least-bad option. Watchman deliberately avoids the use of `/tmp` for security concerns on multi-user systems, since another malicious user could make the watchman tmpdir a symlink. But the last version was effectively usign `/tmp` anyways by setting `--disable-statedir` ([here are the seemingly not-up-to-date docs on the matter](https://facebook.github.io/watchman/docs/cli-options.html#quick-note-on-default-locations))  I wish they would just use something under `$HOME` to mitigate that instead of a hardcoded-at-compile-time path :/

**Comment by andrewhamon** [comment]
It also might be a decent idea to make `WATCHMAN_STATE_DIR` more easily overridable.

**Review by teto — APPROVED** [review]

**Comment by andrewhamon** [comment]
Re: scoping down `patchShebangs`, it seems the relevant files are pretty scattered. A list I arrived at that scoped things down somewhat, while trying to avoid being hyper-specific, is this:

```
autogen.sh
build/fbcode_builder
install-system-packages.sh
run-tests.sh
watchman/build
watchman/integration
watchman/node
watchman/python
watchman/runtests.py
watchman/test
website/publish.sh
```

I think this still touches a significant fraction of files (several hundred at least, out of a total of ~500 files in the watchman repo).

So I'm not sure its worth trying to scope it down unless the size of the watchman repo grows significantly. The more we try to cherry pick things, the more maintenance will be required when upgrading.

**Comment by andrewhamon** [comment]
An alternative way to scope things would be as follows:


```
**/*.py
**/*.sh
watchman/python/bin
```

but thats still >100 files

**Comment by andrewhamon** [comment]
@kylesferrazza is there a particular reason you updated watchman to `fmt_9`? [It looks like watchman still depends on fmt 8](https://github.com/facebook/watchman/blob/72203852145d6b41ef777fa32aa3ec986e89a944/build/fbcode_builder/manifests/fmt) and I think this is the cause of the remaining build failures.

**Comment by kylesferrazza** [comment]
It's been so long I can't remember.. feel free to change it back, thanks.

**Comment by andrewhamon** [comment]
@SuperSandro2000 I think its all ready now. A few follow ups on your prior feedback:

- I changed `WATCHMAN_STATE_DIR` back to `/tmp`.
  - This is functionally equivalent to the prior version (`--disable-statedir` didn't actually disable the statedir, it forced it to use `/tmp`).
  - **This is the only solution I could find that doens't result in a broken-by-default package**.
    - If we use `/run/watchman`, the package isn't usable without the help of `sudo` to first create `/run/watchman`
  - [ I opened an issue upstream about this.](https://github.com/facebook/watchman/issues/1092).
- `patchShebangs` does not seem easy to limit in scope without producing a long and brittle list of paths. The number of files in watchman doesn't seem that large (a few hundred) so I think the most maintainable thing is to leave it scoped as `.`

Aside from that, I believe all your feedback has been taken into account!

**Comment by robstolarz** [comment]
Hi folks! I've been eagerly watching the recent progress on this PR and was really hoping to see it get merged soon. Is there anything I can do to help get it across the line?

**Comment by gsornsen** [comment]
> Hi folks! I've been eagerly watching the recent progress on this PR and was really hoping to see it get merged soon. Is there anything I can do to help get it across the line?

Would like to +1 this and also offer support

**Comment by andrewhamon** [comment]
- rebased to resolve conflicts
- updated everything to `v2023.01.30.00` (to match folly, which was recently upgraded and was the source of the conflicts)
- cleaned up some more cruft (`configureFlags` aren't used anymore since the project moved to CMake)
- made `stateDir` overridable, for any folks not okay with the default
- updated license to mit (watchman [was relicensed](https://github.com/facebook/watchman/commit/bca70f0bfaa2da0de4603ddecf042763f794ee79))

**Comment by andrewhamon** [comment]
@robstolarz @gsornsen testing the package on your platform of choice would be a nice help! I've only tested on m1 macos.

Otherwise, if you know any nixpkgs committers......

@SuperSandro2000 a final review would be much appreciated 🥲

**Comment by andrewhamon** [comment]
@wegank with your merge of https://github.com/NixOS/nixpkgs/pull/204718 folly is now broken on aarch64-linux. This PR includes a fix 😄

**Comment by wegank** [comment]
@andrewhamon That package's broken on aarch64-linux [before my merge](https://hydra.nixos.org/job/nixpkgs/trunk/folly.aarch64-linux).

**Comment by andrewhamon** [comment]
Oh my bad, i assumed the previous version was in a working state (since it was breaking the build in this PR which touched it before your upgrade).

Either way, fixed here.

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 181787` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>buck</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>edencommon</li>
    <li>fb303</li>
    <li>fbthrift</li>
    <li>fizz</li>
    <li>folly</li>
    <li>python310Packages.pywatchman</li>
    <li>python311Packages.pywatchman</li>
    <li>rs-git-fsmonitor</li>
    <li>wangle</li>
    <li>watchman</li>
  </ul>
</details>

**Review by wegank — APPROVED** [review]

**Comment by andrewhamon** [comment]
Thank you to everyone involved!

**Pull request merged**
